plugins/woocommerce/includes/admin/class-wc-admin-notices.php

```diff
@@ -646,9 +646,16 @@ protected static function is_uploads_directory_protected() {
 		$response_code    = intval( wp_remote_retrieve_response_code( $response ) );
 		$response_content = wp_remote_retrieve_body( $response );
 
-		// Check if returns 200 with empty content in case can open an index.html file,
-		// and check for non-200 codes in case the directory is protected.
-		$is_protected = ( 200 === $response_code && empty( $response_content ) ) || ( 200 !== $response_code );
+		// Check that the request either...
+		// 1. Returns a non-200 status code
+		// 2. Has empty or null contents (protected by empty index files; avoids regex test when possible)
+		// 3. Does not look like an Index page (i.e., it has "Index of <...>" in title and contains a <table> element)
+		$is_protected = (
+			200 !== $response_code
+			|| empty( $response_content )
+			|| 1 !== preg_match( '/<title[^>]*>\\s*index\\s+of\\s+[\\s\\S]*<table/i', $response_content )
```

Comment:
My only concern here is that, previously, we may erroneously have flagged that the uploads directory is unprotected when actually it **is** protected. That's not *great,* but at least it prompts the user to go take a look.

With this change, we avoid that issue in some % of cases, but we also risk **not** flagging that there is a problem when a directory listing page not meeting this regex is returned. That doesn't feel like the right trade-off, especially since (at least, following the adoption of the code in the second part of this PR) the admin notice can be dismissed.

Reply:
I mainly added this because in my case I was additionally having this issue, which made it doubly obnoxious. You can refer to my comment on issue #26600 for the reason. But I personally think that issue should either *not* be closed, or should be closed with a 'wontfix' since even the OP references a site for which only something like this code would fix.

I agree that ideally the regex should cover every case. Therefore, I did do research to see if there were other types of index pages from different web engines (e.g., NGINX), and couldn't find any. I also researched to see if a translation or customization would throw off the regex, [but the Apache index page text appears to be hardcoded.](https://stackoverflow.com/questions/27018323/can-i-change-the-language-of-mod-autoindex-apache-directory-listing) So, I left it as is.

But if access to the `woocommerce_uploads` is such a concern, instead of checking if it's accessible or not, why doesn't the plugin just add a blank `index.php` to the folder when the plugin creates it? Wouldn't that be easier?

Reply:
> Since even the OP references a site for which only something like this code would fix

I'm not sure about that, though I may be misunderstanding or something may have changed since that was first posted. They reference [this URL](https://gallmet.hu/wp-content/uploads/woocommerce_uploads) and at time of writing it returns a 403 status—so I'd be surprised if that triggered the warning.

> I did do research to see if there were other types of index pages from different web engines.

Yep, and I appreciate the thorough research you did: I imagine the regex you wrote will cover *most* cases—but it still leaves others unaddressed (Nginx with `autoindex_format json`, various Lighttpd configs, a major non-English language host that has custom compiled Apache2).

Don't misunderstand, I fully understand that is likely a minority of users, but it still feels safer to give a false negative ('something is wrong, go check it out and dismiss this message if all is actually ok') than a false positive ('everything is just fine!' ... but isn't).

> But if access to the woocommerce_uploads is such a concern, instead of checking if it's accessible or not, why doesn't the plugin just add a blank index.php to the folder when the plugin creates it?

It already creates, or tries to create, both an `index.html` file and an `.htaccess` file. An `index.php` file could also be generated, but would require careful consideration first (since that could set off various alarm bells).

Looking at your note in #26599: is `index.html` missing in your case, or, is Nginx configured not to use `.html` files in this role?

Reply:
> They reference [this URL](https://gallmet.hu/wp-content/uploads/woocommerce_uploads) and at time of writing it returns a 403 status—so I'd be surprised if that triggered the warning.

Hmm, yep, you're right! I missed that the status was 403, which would prevent it from being detected.

> Don't misunderstand, I fully understand that is likely a minority of users, but it still feels safer to give a false negative

And yes, I also agree! I was just under the impression that there were a lot more false positives, but...

> It already creates, or tries to create, both an index.html file and an .htaccess file.

Since this is true (I do see them on my server; should've checked first..), then I agree my case is specifically rare. I was under the impression the notice was to encourage the use of `Options -Indexes` without doing something to help prevent the issue. This clearly isn't the case.

> Looking at your note in https://github.com/woocommerce/woocommerce/issues/26599: is index.html missing in your case, or, is Nginx configured not to use .html files in this role?

The specific issue in my case is that I've modified my local system's 'hosts' file (a place for adding custom DNS overrides). Therefore, I'm able to visit the site, even though the domain points elsewhere. But because it points elsewhere, when WooCommerce makes the lookup, it uses the *actual* DNS and gets a different page than I do, which made it really confusing at first.

But if WC does create an index and even .htaccess block, what really would surprise me is finding a true positive case. But if indeed it is possible, then really feel free to leave the second half out. Being able to dismiss the message is certainly an acceptable compromise.

Thanks @barryhughes!

Reply:
> But if WC does create an index and even .htaccess block, what really would surprise me is finding a true positive case.

Yes, though, there are just so many installs out there and such a wide variety of environments with differing file permission issues and web server configs that the possibility is always there.

> Being able to dismiss the message is certainly an acceptable compromise.

Agree—in that case, would you be happy to revert/unwind this change (touching the `base_tables_missing_notice()` method)? I think we can then go ahead and merge the fix for notice dismissal.